src/SourcesCommand.cc

```diff
@@ -130,6 +141,197 @@ struct DirExistsCache {
+  int status;
+  if (waitpid(pid, &status, 0) == -1) {
+    FATAL() << "Failed to wait on gdb script host";
+  }
+
+  if (!(WIFEXITED(status) && WEXITSTATUS(status) == 0)) {
+    FATAL() << "gdb script host in unexpected state " << HEX(status);
+  }
+}
+
+DebugDirManager::DebugDirManager(const string& program, const string& gdb_script)
+  : pipe_fd(-1), pid(-1), output_file(NULL)
+{
+  if (gdb_script.empty()) {
+    return;
+  }
+
+  TempFile output_file = create_temporary_file("rr-gdb-script-host-output-XXXXXX");
+
+  int stdin_pipe_fds[2];
+  if (pipe(stdin_pipe_fds) == -1) {
+    FATAL();
+  }
+
+  posix_spawn_file_actions_t file_actions;
+  int ret = posix_spawn_file_actions_init(&file_actions);
+  if (ret != 0) {
+    FATAL() << "posix_spawn_file_actions_init failed with " << ret;
+  }
+
+  // Close unused write end in the child.
+  ret = posix_spawn_file_actions_addclose(&file_actions, stdin_pipe_fds[1]);
+  if (ret != 0) {
+    FATAL() << "posix_spawn_file_actions_addclose failed with " << ret;
+  }
+
+  // Replace child's stdin with the read end.
+  ret = posix_spawn_file_actions_adddup2(&file_actions, stdin_pipe_fds[0], 0);
+  if (ret != 0) {
+    FATAL() << "posix_spawn_file_actions_adddup2 failed with " << ret;
+  }
+
+  string gdb_script_host_path = resource_path() + "bin/rr-gdb-script-host.py";
+  pid_t pid;
+  char* gdb_script_host_argv[] = {
+    strdup(gdb_script_host_path.c_str()),
+    strdup(output_file.name.c_str()),
+    strdup(gdb_script.c_str()),
+    strdup(program.c_str()),
+    NULL,
+  };
+  ret = posix_spawn(&pid, gdb_script_host_path.c_str(), &file_actions, NULL,
+                    gdb_script_host_argv, environ);
+  if (ret != 0) {
+    FATAL() << "posix_spawn failed with " << ret;
+  }
+
+  // Ignore the return values during cleanup.
+  posix_spawn_file_actions_destroy(&file_actions);
+
+  close(stdin_pipe_fds[0]);
+
+  this->pid = pid;
+  this->pipe_fd = stdin_pipe_fds[1];
+  this->output_file = fopen(output_file.name.c_str(), "r");
+  if (!this->output_file) {
+    FATAL() << "Failed to open gdb script host result file " << output_file.name;
+  }
+
+  synchronize();
+}
+
+DebugDirs DebugDirManager::process_one_binary(const string& binary_path) {
+  if (pipe_fd < 0) {
+    return DebugDirs();
+  }
+
+  auto len = binary_path.length();
+  size_t written = write(pipe_fd, binary_path.c_str(), len);
+  if (written != len) {
+    FATAL() << "Failed to write filename";
+  }
+  written = write(pipe_fd, "\n", 1);
+  if (written != 1) {
+    FATAL() << "Failed to write trailing newline";
+  }
+
+  synchronize();
+  return read_result();
+}
+
+DebugDirs DebugDirManager::read_result() {
+  char buf[4096];
+  DebugDirs result;
+  size_t index;
+  const char delimiter[2] = ":";
+
+  if (pipe_fd < 0) {
+    return result;
+  }
+
+  if (!fgets(buf, sizeof(buf) - 1, output_file)) {
+    FATAL() << "Failed to read gdb script output";
+  }
+  index = strcspn(buf, "\n");
+  buf[index] = 0;
+
+  char* token = strtok(buf, delimiter);
```

Comment:
Are we going to regret not being able to handle paths containing ':'?
I'd prefer an approach that doesn't barf on special characters in file names. Maybe \n is already a lost cause, but other than that...

Reply:
Using the path separator here comes from gdb, not from me. Using colons in directory names in Linux is already heavily frowned upon because you can't e.g. add them to PATH. I don't think we'll regret this.